### ISSUE TITLE
Expose `vim.*` to uv threads

### ISSUE THREAD

**Issue by lewis6991** [neovim/neovim#15655]
### Feature already in Vim?

No

## Problem

Libuv provides some pretty cool [ways](https://github.com/luvit/luv/blob/master/docs.md#thread-pool-work-scheduling) to enable multithreading, however they have some pretty big limitations:

1. They only support passing simple [types](https://github.com/luvit/luv/blob/master/docs.md#pseudo-types) (`nil`, `number`, `boolean`, `string`, `userdata`) to/from the main thread.
2. None of the nvim-lua globals (`vim.*`) are available within the thread.

## Solution

Expose all/some of `vim.*` to uv threads.

At a minimum we should be able to expose: [mpack](https://github.com/neovim/neovim/pull/15566), [xdiff](https://github.com/neovim/neovim/pull/14536) and [cjson (TBD)](https://github.com/neovim/neovim/pull/14871). mpack will resolve 1) since it will allow us to serialize/deseralize more complex types.

## Implementation

One way this could be done could be by providing a `vim.so` lib which can then be loaded (via `require`) in a thread.

## Example

```lua
local a = ...;
local b = ...;

vim.loop.new_work(
  function(a0, b0)
     local vim = require('vim')
     return vim.diff(a0, b0)
  end,
  function(r)
    [use diff result]
  end
):queue(a, b)

```

I can probably handle developing the lib itself, but would need some help with the cmake plumbing to actually build the so file.

**Comment by bfredl** [comment]
we should use `luv_set_thread_cb` for this. This lets us initialize the shared parts of the `vim.*` API in a callback.

**Comment by bfredl** [comment]
It would be a tricky as the vim regex engine likes to do breakcheck. But we need to fix it anyway to be able to use regex inside of luv callbacks.

**Comment by hrsh7th** [comment]
Thank you for kindful response. I've understand. it's difficult .

**Comment by ray-x** [comment]
Agree. Some of the APIs  (even some of vim.api.*)  can not be executed in the thread context.
